File: AdventOfCode2022/Day05/day05.js
const fs = require("fs");
const data = fs.readFileSync("Day05/input.txt", "utf8");
const lines = data.split("\n").map((x) => x.trim());

let stacksInput = [
  "",
  "QWPSZRHD",
  "VBRWQHF",
  "CVSH",
  "HFG",
  "PGJBZ",
  "QTJHWFL",
  "ZTWDLVJN",
  "DTZCJGHF",
  "WPVMBH",
].map((x) => x.split(""));
let i = 0;
while (lines[i]) {
  i++;
}

i++;

let moves = [];
for (; i < lines.length; i++) {
  const split = lines[i].split(" ");
  const quantity = +split[1];
  const from = +split[3];
  const to = +split[5];
  moves.push({ quantity, from, to });
}

// part 1
const moveByOne = (stacks, move) => {
  for (let k = 0; k < move.quantity; k++) {
    const crate = stacks[move.from].pop();
    stacks[move.to].push(crate);
  }
};

const stacks1 = stacksInput.map((x) => [...x]);
for (let i = 0; i < moves.length; i++) {
  moveByOne(stacks1, moves[i]);
}

const part1 = stacks1.map((x) => x[x.length - 1]).join("");
console.log(part1);


// part 2
const moveByAll = (stacks, move) => {
  const crates = stacks[move.from].splice(stacks[move.from].length - move.quantity);
  stacks[move.to] = [...stacks[move.to], ...crates];
};

const stacks2 = stacksInput.map((x) => [...x]);
for (let i = 0; i < moves.length; i++) {
  moveByAll(stacks2, moves[i]);
}

const part2 = stacks2.map((x) => x[x.length - 1]).join("");
console.log(part2);

File: AdventOfCode2022/Day05/Day05.hs
import Data.List.Split (splitWhen)

stacks =
  [ "QWPSZRHD",
    "VBRWQHF",
    "CVSH",
    "HFG",
    "PGJBZ",
    "QTJHWFL",
    "ZTWDLVJN",
    "DTZCJGHF",
    "WPVMBH"
  ]

parseMove s =
  let [_, quantity, _, from, _, to] = words s
   in (read quantity, read from, read to) :: (Int, Int, Int)

move stacks (quantity, from, to)
  | quantity == 0 = stacks
  | otherwise = move newStacks (quantity - 1, from, to)
  where
    crate = last . snd $ stacks !! (from - 1)
    newStacks = map moveCrate stacks
    moveCrate (i, stack)
      | i == from = (i, init stack)
      | i == to = (i, stack ++ [crate])
      | otherwise = (i, stack)

move2 stacks (quantity, from, to) = newStacks
  where
    (_, fromStack) = stacks !! (from - 1)
    fromNewLength = length fromStack - quantity
    crate = drop fromNewLength fromStack
    newStacks = map moveCrates stacks
    moveCrates (i, stack)
      | i == from = (i, take fromNewLength stack)
      | i == to = (i, stack ++ crate)
      | otherwise = (i, stack)

main = do
  content <- readFile "input.txt"
  let movesStr = last . splitWhen (== "") $ lines content
  let moves = map parseMove movesStr
  let stacksT = zip [1,2 .. length stacks] stacks
  let part1 = foldl move stacksT moves
  print $ map (last . snd) part1

  let part2 = foldl move2 stacksT moves  
  print $ map (last . snd) part2
